src/test/scala/mesosphere/marathon/api/validation/AppDefinitionValidatorTest.scala

```diff
@@ -452,6 +452,23 @@ class AppDefinitionValidatorTest extends MarathonSpec with Matchers with GivenWh
     AppDefinition.validAppDefinition(to3).isSuccess should be(true)
   }
 
+  test("Resident app may not define acceptedResourceRoles") {
+    Given("A resident app definition")
+    val f = new Fixture
+    val from = f.validResident
+    AllConf.SuppliedOptionNames = Set("mesos_authentication_principal", "mesos_role", "mesos_authentication_secret_file")
+
+    When("validating with acceptedResourceRoles")
+    val to1 = from.copy(acceptedResourceRoles = Some(Set("foo")))
+    Then("Should be invalid")
+    AppDefinition.validAppDefinition(to1).isSuccess shouldBe false
+
+    When("validating without acceptedResourceRoles")
+    val to2 = from.copy(acceptedResourceRoles = None)
```

Comment:
Maybe, I misunderstand something. I completely understand that only checking whether `acceptedResourceRoles` is set, is not sufficient. You basically have to check the "effective" acceptedResourceRoles. And the effective `acceptedResourceRoles` needs to be `Set("*")`.

Your code ignores this user setting. That means that the user said it that the task should only be run on a specific role (by the global setting since you disallow the other) and you explicitly ignore this. That is not good either.

Furthermore, we should allow `Set("*")` since it pretty much expresses what we do.

Since the global default may change, you can have an additional safe-guard at the place of making the reservation which checks whether we attempt to match resources other than "*".

Does that make sense?